src/sql/src/query_model/lowering.rs

```diff
@@ -0,0 +1,222 @@
+                let correlation_info = the_box.correlation_info(&self.model);
+                if !correlation_info.is_empty() {
+                    panic!("correlated joins are not supported yet");
+                }
+
+                let outer_arity = get_outer.arity();
+                let (mut input, column_map) =
+                    self.lower_join(get_outer, outer_column_map, &the_box.quantifiers);
+                let input_arity = input.arity();
+
+                // 2) Lower the filter component.
+                if !select.predicates.is_empty() {
+                    input = input.filter(
+                        select
+                            .predicates
+                            .iter()
+                            .map(|p| Self::lower_expression(p, &column_map)),
+                    );
+                }
+
+                // 3) Lower the project component.
+                if !the_box.columns.is_empty() {
+                    input = input.map(
+                        the_box
+                            .columns
+                            .iter()
+                            .map(|c| Self::lower_expression(&c.expr, &column_map))
+                            .collect_vec(),
+                    );
+                }
+
+                // Project the outer columns plus the ones in the projection of
+                // this select box
+                input.project(
+                    (0..outer_arity)
+                        .chain(input_arity..input_arity + the_box.columns.len())
+                        .collect_vec(),
+                )
+            }
+            _ => panic!(),
+        }
+    }
+
+    /// Generates a join among the result of applying the outer relation to the given
+    /// quantifiers. Returns a relation and a map of column references that can be
+    /// used to lower expressions that sit directly on top of the join.
+    ///
+    /// The quantifiers are joined on the columns of the outer relation.
+    /// TODO(asenac) Since decorrelation is not yet supported the outer relation is
+    /// currently always the join identity, so the result of this method is always
+    /// a cross-join of the given quantifiers, which makes part of this code untesteable
+    /// at the moment.
+    fn lower_join(
+        &mut self,
+        get_outer: expr::MirRelationExpr,
+        outer_column_map: &ColumnMap,
+        quantifiers: &QuantifierSet,
+    ) -> (expr::MirRelationExpr, ColumnMap) {
+        let outer_arity = get_outer.arity();
+
+        // TODO(asenac) Lower scalar subquery quantifiers
+        assert!(quantifiers.iter().all(|q_id| {
+            self.model.get_quantifier(*q_id).borrow().quantifier_type == QuantifierType::Foreach
+        }));
+
+        let join_inputs = quantifiers
+            .iter()
+            .map(|q_id| {
+                let input_box = self.model.get_quantifier(*q_id).borrow().input_box;
+                self.apply(input_box, get_outer.clone(), outer_column_map)
```

Comment:
Maybe add a `@todo` here to reuse the `outer` with `let` and `get` if the number of `quantifiers.length() > 1`.

Reply:
`get_outer` is meant to be a `Get` already, hence its name. I'll add an assert.